### PR TITLE
Reduce the frequency of ornaments calling layoutSubviews

### PR DESCRIPTION
**Pull request by jmkiley**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-ios/blob/main/CONTRIBUTING.md#contributor-license-agreement).

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Add an entry inside this element for inclusion in the `mapbox-maps-ios` changelog: `<changelog></changelog>`.
 - [ ] Update the migration guide, API Docs, Markdown files - Readme, Developing, etc

### Summary of changes

This PR reduces how frequently `layoutSubviews` is called by the map's ornaments, primarily the scale bar, during camera changes. This is done by:
- Checking whether the compass view's old value for `currentBearing` is equal to the new value before updating the compass.
- Place the compass image within a container `UIImageView`.
- Moved the scale bar into a container UIView and moved the override code from `updateConstraints` to a new `updateScaleBarConstraints` method.

I do still see occasional calls to `layoutSubviews()`, but those seem to come from `BaseMapView`.

### CODE REVIEW

File: Apps/DebugApp/DebugApp/DebugViewController.swift [hunk 1]
```diff
@@ -90,4 +91,18 @@ public class DebugViewController: UIViewController {
             print("The map failed to load.. \(type) = \(message)")
         }
     }
+
```
Comment (neelmistry94) [inline Apps/DebugApp/DebugApp/DebugViewController.swift:94]: I feel like we can remove this debug code

File: Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift [hunk 2]
```diff
@@ -86,7 +86,7 @@ internal class MapboxScaleBarOrnamentView: UIView {
 
     private var labelImageCache: [CLLocationDistance: UIImage] = [:]
     private var lastLabelWidth: CGFloat = Constants.scaleBarLabelWidthHint
-    private var size = CGSize()
+    var size: CGSize = CGSize()
```
Comment (neelmistry94) [inline Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift:89]: You don't need the `: CGSize` that is implied when you are creating it

File: Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift [hunk 3]
```diff
@@ -31,17 +31,19 @@ internal class MapboxCompassOrnamentView: UIButton {
...
-            updateVisibility()
-            self.transform = CGAffineTransform(rotationAngle: -adjustedBearing.toRadians())
+            if oldValue != currentBearing {
+                let adjustedBearing = currentBearing.truncatingRemainder(dividingBy: 360)
+                updateVisibility()
+                self.compassContainerView.transform = CGAffineTransform(rotationAngle: -adjustedBearing.toRadians())
+            }
         }
     }
+    internal var compassContainerView = UIImageView()
```
Comment (neelmistry94) [inline Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift:41]: Nit: newline

File: Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift [hunk 4]
```diff
@@ -212,7 +214,7 @@ internal class MapboxScaleBarOrnamentView: UIView {
         let barOffset = isRightToLeft ? halfLabelWidth : 0.0
 
         containerView.frame = CGRect(x: barOffset,
-                                     y: intrinsicContentSize.height - Constants.barHeight,
+                                     y: 20 - Constants.barHeight,
```
Comment (neelmistry94) [inline Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift:217]: What is the purpose of this arbitrary value of 20? Can we add docs around this

File: Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift [hunk 5]
```diff
@@ -55,7 +57,11 @@ internal class MapboxCompassOrnamentView: UIButton {
                                               comment: "Accessibility hint")
 
         if let image = createCompassImage() {
+            let bounds = CGRect(origin: .zero, size: image.size)
+            compassContainerView.frame = bounds
             setImage(image, for: .normal)
+            compassContainerView.alpha = visibility == .visible ? 1 : 0
```
Comment (julianrex) [inline Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift:63]: does this mean `adaptive` is not visible? Does that clash with any docs?

File: Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift [hunk 6]
```diff
@@ -55,7 +57,11 @@ internal class MapboxCompassOrnamentView: UIButton {
...
+            self.addSubview(compassContainerView)
```
Comment (julianrex) [inline Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift:64]: nit: let's avoid self where we can.

File: Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift [hunk 7]
```diff
@@ -86,7 +86,7 @@ internal class MapboxScaleBarOrnamentView: UIView {
 
     private var labelImageCache: [CLLocationDistance: UIImage] = [:]
     private var lastLabelWidth: CGFloat = Constants.scaleBarLabelWidthHint
-    private var size = CGSize()
+    var size = CGSize()
```
Comment (julianrex) [inline Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift:89]: Why not private?

File: Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift [hunk 8]
```diff
@@ -144,14 +150,13 @@ internal class MapboxScaleBarOrnamentView: UIView {
...
     // MARK: - Layout
 
-    // The primary job of 'updateConstraints' here is to recalculate the
+    // The primary job of 'updateScaleBarConstraints' here is to recalculate the
     // 'intrinsicContentSize:', 'metersPerPoint' and the maximum width determine the
     // current 'row', which in turn determines the "actualWidth". To obtain the full
     // width of the scale bar, we also need to include some space for the "last"
     // label
-    internal override func updateConstraints() {
+    internal func updateScaleBarConstraints() {
```
Comment (julianrex) [inline Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift:158]: Let's replace "constraints" if we're modifying them

File: Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift [hunk 9]
```diff
@@ -180,9 +184,7 @@ internal class MapboxScaleBarOrnamentView: UIView {
         let halfLabelWidth = ceil(lastLabelWidth / 2)
 
         size = CGSize(width: totalBarWidth + halfLabelWidth, height: 16)
```
Comment (julianrex) [inline Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift:186]: So this is modifying size and calling setNeedsLayout. Although we're not calling updateConstraints, `intrinsicContentSize` is returning size - so I wonder what the behavior is if there's a view to the side of the scale bar.

How does the scale bar look when it's on the right hand side of the view?

File: Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift (hunk 5 above)
Comment (jmkiley) [reply Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift:63]: This is adapted from the [previous way](https://github.com/mapbox/mapbox-maps-ios/blob/main/Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift#L44) we set the alpha for the compass view, just applied to the compass container view, so behavior shouldn't change.

File: Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift (hunk 9 above)
Comment (jmkiley) [reply Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift:185]: I am working on resolving some issues when the scale bar is on the top right. Currently only seeing the `0`. If I restore `updateScaleBar` to `updateConstraints`, I see the scaleBar but `layoutSubviews` is called repeatedly.

File: Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift [hunk 10]
```diff
@@ -43,25 +48,26 @@ internal class MapboxScaleBarOrnamentView: UIView {
...
     }
 
-    lazy private var containerView: UIView = {
+    var size = CGSize()
+    // This container view's size and position can change based on the size
+    // of its contents. It is contained within the `staticContainerView`.
+    lazy private var dynamicContainerView: UIView = {
         let view = UIView()
-        view.clipsToBounds = true
+//        view.clipsToBounds = true
```
Comment (neelmistry94) [inline Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift:64]: we shouldn't use commented out code on the final commit

File: Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift [hunk 11]
```diff
@@ -131,9 +133,29 @@ internal class MapboxScaleBarOrnamentView: UIView {
...
+        // Add width constraint to ornament view
+        // Pin scale bar container to match root ornament view
+        // Test against all four configurations
+
+//        staticContainerView.leftAnchor.constraint(equalTo: safeAreaLayoutGuide.leftAnchor).isActive = true
+        staticContainerView.widthAnchor.constraint(equalTo: safeAreaLayoutGuide.widthAnchor).isActive = true
+        staticContainerView.heightAnchor.constraint(equalTo: safeAreaLayoutGuide.heightAnchor).isActive = true
+        staticContainerView.topAnchor.constraint(equalTo: safeAreaLayoutGuide.topAnchor).isActive = true
+        staticContainerView.rightAnchor.constraint(equalTo: safeAreaLayoutGuide.rightAnchor).isActive = true
+//        staticContainerView.bottomAnchor.constraint(equalTo: safeAreaLayoutGuide.bottomAnchor).isActive = true
```
Comment (neelmistry94) [inline Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift:155]: same comments above, remove commented out code

File: Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift (hunk 10 above)
Comment (jmkiley) [reply Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift:64]: Agreed, I'm still debugging this code.

**Comment by jmkiley** [comment]
Working through a couple of constraints issues when the scale bar is in positions other than `.topLeft`.

**Comment by jmkiley** [comment]
@julianrex @neelmistry94  I think that this is ready for re-review when you have a chance.

File: Sources/MapboxMaps/Foundation/Camera/BasicCameraAnimator.swift [hunk 12]
```diff
@@ -78,7 +78,7 @@ public class BasicCameraAnimator: NSObject, CameraAnimator, CameraAnimatorInterf
             }
 
             // Set up the short lived camera view
-            delegate.addViewToViewHeirarchy(cameraView)
+//            delegate.addViewToViewHeirarchy(cameraView)
```
Comment (neelmistry94) [inline Sources/MapboxMaps/Foundation/Camera/BasicCameraAnimator.swift:81]: commented out code here, these comments need to be removed

File: Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift [hunk 13]
```diff
@@ -33,15 +34,15 @@ internal class MapboxCompassOrnamentView: UIButton {
...
+            self.containerView.transform = CGAffineTransform(rotationAngle: -adjustedBearing.toRadians())
         }
     }
 
     required internal init(visibility: OrnamentVisibility) {
         super.init(frame: .zero)
         translatesAutoresizingMaskIntoConstraints = false
         self.visibility = visibility
-        alpha = visibility == .visible ? 1 : 0
+        containerView.alpha = visibility == .visible ? 1 : 0
```
Comment (neelmistry94) [inline Sources/MapboxMaps/Ornaments/Compass/MapboxCompassOrnamentView.swift:45]: if `containerView` is a `UIView` we should use the `isHidden` property. Messing with alphas for visibility is not right

File: Sources/MapboxMaps/Ornaments/OrnamentsManager.swift [hunk 14]
```diff
@@ -40,6 +40,7 @@ public class OrnamentsManager: NSObject {
 
         // Scalebar View
         scalebarView = MapboxScaleBarOrnamentView()
+        scalebarView.isOnRight = options.scaleBar.position == OrnamentPosition.bottomRight || options.scaleBar.position == OrnamentPosition.topRight
```
Comment (neelmistry94) [inline Sources/MapboxMaps/Ornaments/OrnamentsManager.swift:43]: this is confusing to read. It makes sense, but very complex for a ternary case. Can this be put in an actual if statement

File: Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift [hunk 15]
```diff
@@ -221,13 +236,14 @@ internal class MapboxScaleBarOrnamentView: UIView {
         let barDelta = isRightToLeft ? -barWidth : barWidth
 
         layoutLabels(with: barOffset, delta: barDelta, yPosition: yPosition)
+
```
Comment (neelmistry94) [inline Sources/MapboxMaps/Ornaments/ScaleBar/MapboxScaleBarOrnamentView.swift:239]: nit: newline

**Review by neelmistry94 — CHANGES_REQUESTED** [review]

**Review by neelmistry94 — APPROVED** [review]

**Pull request merged**